### ISSUE TITLE
Package verdict is match, but percentage isn't 100%?

### ISSUE THREAD

**Issue by peterjaap** [sansecio/composer-integrity-plugin#8]
```
|   ⨉    | catgento/module-admin-activity                           | 1.1.5       | A4F12DC1F99F950F | F7276DCEDE410FC8 |    29%     |      Yes       |
|   ✓    | community-engineering/language-nl_di                     | 0.0.20      | 67661929CB2DD969 | 52A49ACA7C92CA5B |    74%     |      Yes       |
|   ⨉    | magento/framework                                        | 103.0.6     | A84A3C75B09672D6 | B27E7A2573F62A97 |     0%     |      Yes       |
|   ?    | magento/magento-coding-standard                          | 31          | 883309F3850B223A | F1C5AF4221B5FF1D |     -      |      Yes       |
|   ⨉    | magento/magento2-base                                    | 2.4.6       | DE99F811913B97C2 | D01FF0BFB3B4AFD1 |     4%     |      Yes       |
|   ⨉    | magento/module-catalog                                   | 104.0.6     | 1A9765C9211C614E | 14C22B707B00D951 |     0%     |      Yes       |
|   ⨉    | magento/module-sales                                     | 103.0.6     | 40AC78C97F30F0ED | 542480E720A16198 |     2%     |      Yes       |
|   ⨉    | mageplaza/module-webhook                                 | 4.1.3       | 690F61FFDEAD6D13 | AC38792030BBD1C7 |    29%     |      Yes       |
|   ✓    | prismic/php-sdk                                          | 5.1.3       | 038E4F3537053F89 | 9F2626447E0202A6 |    100%    |      Yes       |
|   ✓    | smile/elasticsuite                                       | 2.11.1      | C6A6AE20F7E32850 | C52FB223FCCE25E9 |    53%     |      Yes       |
|   ⨉    | yireo/magento2-googletagmanager2                         | 2.1.8       | 824AC63E2C724D9A | 5341B76FAA51D6B9 |     4%     |      Yes       |

```

Given this output, why is the verdict for `smile/elasticsuite` (for example) a `match`, even though we have a patch for it installed (which means the code differs locally)  _and_ the incidence percentage is only 53%?

**Comment by danslo** [comment]
As for why it's passing with only 53% incidence rate, is because there are a lot of modules that are patched in a certain way. This means they are not the same as a clean composer install, but should still not be flagged as mismatching.
An example is those MDVA hotfixes released by Magento.

I'm not exactly sure what the tresholds are (@gwillem ), but I believe it was something like >50 reports with >50% (or even >30%) is still okay. We are able to tweak these on the serverside, we'll need to collect a bit more data to see where the sweet spot is :)

**Comment by peterjaap** [comment]
@danslo aaah thanks, that makes sense, although that would mean that for `magento/module-catalog` 100% of all installs have a patch applied? I do find that hard to believe.

**Comment by peterjaap** [comment]
So looking at this line;

```
+--------+-------------------------------------------------+-------------+------------------+------------------+------------+----------------+
| Status | Package                                         | Version     | Package ID       | Checksum         | Percentage | Patch applied? |
+--------+-------------------------------------------------+-------------+------------------+------------------+------------+----------------+
|   ⨉    | mageplaza/module-webhook                        | 4.1.3       | 690F61FFDEAD6D13 | AC38792030BBD1C7 |    29%     |      Yes       |
```

This 29% means that out of all the installs you have seen on your end, 29% match with the checksum that is generated based on my local code (which includes a patch)?

**Comment by danslo** [comment]
> @danslo aaah thanks, that makes sense, although that would mean that for `magento/module-catalog` 100% of all installs have a patch applied? I do find that hard to believe.

No it means that the hash you sent us of that package doesn't match anyone else's hash. I'd be interested to see the contents :)

In the case of `mageplaza/module-webhook`, 29% of the checksums in our database for `mageplaza/module-webhook:4.1.3` are the same as yours. This could be due to manual patches, or maybe mageplaza released 2 different packages under the same version, or there was a MDVA hotfix, etc.

Or maybe we just didn't gather enough data yet for this percentage to be accurate (this "minimum checksums gathered" 
 is also a number that we can tweak).

**Comment by peterjaap** [comment]
Here is the zipped content of that module: [magento-module-catalog.zip](https://github.com/sansecio/composer-integrity-plugin/files/11309005/magento-module-catalog.zip)

**Comment by danslo** [comment]
> magento/module-catalog

```
$ curl -H "User-Agent: Composer/2.0.7 (Linux; 4.15.0-20-generic; PHP 7.2.24; curl 7.58.0)" --user USERNAME:PASSWORD https://repo.magento.com/archives/magento/module-catalog/magento-module-catalog-104.0.6.0.zip -o catalog.zip
$ unzip catalog.zip
```

```
diff -qr ~/Downloads/vendor\ 2/magento/module-catalog/ /tmp/catalog
Files /Users/daniel/Downloads/vendor 2/magento/module-catalog/Block/Breadcrumbs.php and /tmp/catalog/Block/Breadcrumbs.php differ
Files /Users/daniel/Downloads/vendor 2/magento/module-catalog/view/adminhtml/templates/catalog/product/attribute/js.phtml and /tmp/catalog/view/adminhtml/templates/catalog/product/attribute/js.phtml differ
```

```
diff "/Users/daniel/Downloads/vendor 2/magento/module-catalog/Block/Breadcrumbs.php" /tmp/catalog/Block/Breadcrumbs.php
59,60d58
<         $path = $this->_catalogData->getBreadcrumbPath();
<
69a68,71
>
>             $title = [];
>             $path = $this->_catalogData->getBreadcrumbPath();
>
71a74
>                 $title[] = $breadcrumb['label'];
73d75
<         }
75,76c77,78
<         $title = array_column($path, 'label');
<         $this->pageConfig->getTitle()->set(join($this->getTitleSeparator(), array_reverse($title)));
---
>             $this->pageConfig->getTitle()->set(join($this->getTitleSeparator(), array_reverse($title)));
>         }
```

```
diff "/Users/daniel/Downloads/vendor 2/magento/module-catalog/view/adminhtml/templates/catalog/product/attribute/js.phtml" "/tmp/catalog/view/adminhtml/templates/catalog/product/attribute/js.phtml"
216c216
< foreach ($this->helper(Magento\Catalog\Helper\Data::class)->getAttributeHiddenFields() as $type => $fields):
---
> foreach ($jsonHelper->getAttributeHiddenFields() as $type => $fields):
```

The files differ from repo.magento.com

**Comment by peterjaap** [comment]
We use the Mage-OS mirror for this particular project. Apparently it has different code for the same versions?

**Comment by danslo** [comment]
```
/tmp/magento$ curl -H "User-Agent: Composer/2.0.7 (Linux; 4.15.0-20-generic; PHP 7.2.24; curl 7.58.0)" --user USER:PASSWORD https://repo.magento.com/archives/magento/module-catalog/magento-module-catalog-104.0.6.0.zip -o catalog.zip
/tmp/magento$ unzip catalog.zip && rm catalog.zip
```

```
/tmp/mageos$ curl https://mirror.mage-os.org/packages/magento/module-catalog-104.0.6.zip -o catalog.zip
/tmp/mageos$ unzip catalog.zip && rm catalog.zip
```

```
diff -qr /tmp/mageos /tmp/magento
Files /tmp/mageos/composer.json and /tmp/magento/composer.json differ
```

They do differ, but not in the files you mentioned. The hasher only considers php/html/phtml/js files.

**Comment by peterjaap** [comment]
@danslo omg, I was stupid.

I ran `composer install` for this particular project with `--no-plugins` so `vaimo/composer-patches` wouldn't fire.

However, we also have this in this projects' `composer.json`;

```
    "scripts": {
        "post-install-cmd": [
            "./vendor/bin/magento-patches status | grep 'Not applied' | cut -d ' ' -f2 | xargs --no-run-if-empty ./vendor/bin/magento-patches apply"
        ]
    },
```

This automatically applies all [quality patches](https://github.com/magento/quality-patches) and these were still applied (silently). So I ran `composer install --no-scripts --no-plugins` followed by `composer integrity` and now all the Magento module mentions are obviously gone...

Woops! Closing :)
